[JENKINS-33364] Core documentation: Agent configuration

orrc opened the pull request:
Ticket: https://issues.jenkins-ci.org/browse/JENKINS-33364
This PR covers: Agent configuration

Related PRs: #2136, #2137

---

I'm going through Jenkins core on the 2.0 branch (without any plugins), and fixing up documentation to be more consistent, removing unnecessary detail, fixing broken English, replacing stray instances of "slave" with "agent", and adding documentation where it's missing entirely.

---

My mini "styleguide" of terms so far (open to discussion):

| Canonical term | Term(s) to be replaced | Comments |
| --- | --- | --- |
| an agent | a agent, slave | There are stray usages of the old term, and search-replace errors where "a" instead of "an" is used |
| for example | for instance | There _is_ a difference, but it's unlikely that we will need "for instance" |
| hostname | host name, host-name | [Most common usage](https://en.wikipedia.org/wiki/Hostname) |
| label expression | label restriction | The job config field is named "Label expression" |
| log in | login (verb) | Login is **not** a verb; it's ok when used as a noun |
| master | Master | Keep it lowercase |
| offline | off-line |  |
| online | on-line |  |
| project | job |  |
| use | utilize | Keep it simple |
| does not | doesn't |  |
| it is | it's |  |
- Field names should use sentence case (i.e. "Keep this agent online" rather than "Keep this Agent Online").
- When referring to other field names in documentation, use italics — e.g. `This depends on the <i>Launch method</i> selected.`
- Use "builds" rather than "jobs" as appropriate.
- To avoid interrupting config editing, links in documentation should open in a new tab, using the `target="_blank"` attribute. e.g. `<a href="https://en.wikipedia.org/wiki/JNLP" target="_blank">...</a>`.

orrc committed:
Update documentation for the Agent configuration page.

orrc commented on core/src/main/resources/hudson/model/Node/help-labelString.html:
```diff
@@ -1,11 +1,22 @@
+  <p>
+  Labels may contain any non-space characters, but you should avoid special
+  characters such as any of these: <tt>!&|<>()</tt>, as other Jenkins features
+  allow for defining label expressions, where these characters may be used.
```
Yes, you can have labels called `||` and `&&` — surprisingly this works: `"||" || "&&"` is a valid label expression.

However, `<blink/>` is accepted but causes a crash.  Should file an issue for that.  Or we should apply the same "is safe" check to label names as we do to agent name.s

batmat replied:
Shouldn't those \<> and so on be htmlentities encoded? I think so at least for \<>, and probably for & too (GH makes them going red btw, nice, why I saw that)

orrc commented on core/src/main/resources/hudson/model/Node/help-name.html:
```diff
@@ -1,7 +1,9 @@
+  but it is often convenient to make them the same.
+  <p>
+  The name may not contain any characters from this list:
+  <tt>?*/\\%!@#$^&|<>[]:;</tt>
```
I know these characters should generally be escaped, but Jenkins works fine without (due to being in a `<tt>` tag?).

batmat replied:
right. /me think this would still preferable in general as IMO it's a browser matter and we're not sure it's gonna render fine everywhere (or less sure I guess).

orrc replied:
Yeah, probably best to fix this.

daniel-beck commented on core/src/main/resources/hudson/model/Node/help-labelString.html:
```diff
@@ -1,11 +1,22 @@
+  must run on Windows, then you can configure all your Windows agents to have
+  the label <tt>windows</tt>, and then tie that job to this label.
+  <br>
+  This would ensure that your job runs on one of your Windows agents, but not
```
*This ensures*

orrc replied:
Yeah, sure. Tenses are still a bit mixed up here anyway.

Also, I wonder about:
- *your job runs* on of your agents
- *the job gets built* on one of your agents
- *builds of this job* run on one of your agents

etc.

daniel-beck commented on core/src/main/resources/hudson/model/Node/help-labelString.html:
```diff
@@ -1,11 +1,22 @@
+  Labels (or tags) are used to group multiple agents into one logical group.
+  <p>
+  For example, if you have multiple Windows agents and you have a job that
+  must run on Windows, then you can configure all your Windows agents to have
```
+ *and want Jenkins to handle assignment and load distribution*

orrc replied:
I'm not sure this adds anything.

daniel-beck commented on core/src/main/resources/hudson/model/Node/help-labelString.html:
```diff
@@ -1,11 +1,22 @@
+  you can also use labels to note the CPU architecture, or that a certain tool
+  is installed on the agent.
+  <p>
+  Multiple labels should be separated by a space. For example,
```
*must*

orrc replied:
I'm not sure.  "Should" reads better to me — this isn't an RFC with `MUST` and `SHOULD` etc.

But strictly, yeah, it is a "must".  This is something we should have in the styleguide.

orrc replied:
Thinking about it some more, I think `must` is good here.

jtnord replied:
must is a little aggressive, consider "are separated by" instead?

daniel-beck commented on core/src/main/resources/hudson/model/Node/help-numExecutors.html:
```diff
@@ -1,15 +1,12 @@
-  When using Jenkins in the master/agent mode, setting this value to 0 would prevent the master
-  from doing any building on its own.  Agents may not have zero executors, but may be
-  temporarily disabled using the button on the agent's status page.
+  Setting this value to 0 will prevent the master from executing any builds.
```
Isn't this help file shared between master and agents, making the previous paragraph much more informative?

orrc replied:
As far as I could tell at the time, it's not.  Part of it certainly could / should be.

orrc replied:
Yeah, looks like it might be shared after all. I'll need to check if there is indeed another similar file (I thought there was), and fix this up again.

daniel-beck commented on war/src/main/webapp/help/system-config/master-slave/usage.html:
```diff
@@ -6,7 +6,8 @@
       Utilize this node as much as possible
```
Left-over *utilize*

jtnord replied:
@daniel-beck what does your comment mean?

daniel-beck replied:
The initial PR comment listed *utilize* to be replaced by *use*. In this particular case it would require more than just changing the documentation, so may be out of scope.

orrc replied:
Yeah, I updated the table after submitting the PR :)

Also, I hadn't changed this yet, as it has a dependency on a `.properties` translation.

I also wanted to ask in general, if we can make this documentation dynamic, since this help text depends on what's in the launchers drop-down.

orrc replied:
Ah, I was confusing this ("Usage") with "Launch method", which is already dynamic.

I will change this to "Use", as on the code side it is just a very simple `.properties` change.

jtnord commented on war/src/main/webapp/help/system-config/master-slave/usage.html:
```diff
@@ -19,11 +20,11 @@
       and that expression matches this node&apos;s name and/or labels.
 
-      This allows a slave to be reserved for certain kinds of jobs.
+      This allows a node to be reserved for certain kinds of jobs.
```
node agent?  confused I am....

daniel-beck replied:
It applies to both master and agents (formerly slaves). Master is not an agent, but a node.

orrc replied:
@jtnord Yes, I wish we didn't have to have this inconsistency :disappointed: 

I kept the wording here as "node" because the launcher mode is currently called "Utilize this **node** as much as possible", and the rest of the text talks about "nodes" — even although we're on the *agent* configuration page, and all the other inline documentation here talks about "agents".

I opened [JENKINS-33007](https://issues.jenkins-ci.org/browse/JENKINS-33007) recently; I believe all instances of "node" should be replaced with "agent".  Just because internally everything is called a `Node`, does not mean that users should be subjected to this terminology.  Even before the "slave" -> "agent" rename, this could have been better, but it makes more sense now that slaves have a more descriptive name.

The Jenkins master server has two roles: it handles configuration, queue management etc.  It also can optionally act as an *agent*, executing builds, much like any other SSH or JNLP or whatever agent.

So in 2.0 we currently end up with this delightful mixture of terms:
![agent-node-mixture](https://cloud.githubusercontent.com/assets/138615/13638057/d8da961a-e60a-11e5-9d34-dddd4a500fdb.png)

I know @daniel-beck wants to keep "node", but I haven't been able to understand why.  I really think JENKINS-33007 would be a beneficial change for consistency and clarity, and therefore end users, but for this PR I'm sticking to what fits with the immediate context (i.e. Utilize this node).

rtyler commented:
I just want to say, on behalf of confused Jenkins users everywhere, thanks for driving this @orrc :smiley_cat: 

I hope to see it lose the "WIP" tag soonish

daniel-beck commented:
@orrc We're trying to get the beta out on Wednesday. Any change this is done by then?

orrc commented:
@daniel-beck Nope, no chance! :) I've had no time this week, and will have none tomorrow. :disappointed:

rtyler commented:
@daniel-beck perhaps it's worth trying to find some other sucker^Wnice person to do this?

@omehegan what are you up to this week? :smiling_imp:

orrc commented:
I found a wee bit of time today... next week I should have lots of time.

Anyway, as this is a pretty big task overall, I've split it into several PRs.  I will try and complete the others in the priority order I set in [JENKINS-33364](https://issues.jenkins-ci.org/browse/JENKINS-33364).

This PR is near-complete: I broke and still need to fix `help-numExecutors.html`, as mentioned by Daniel in the comments above.

daniel-beck commented:
> Field names should use sentence case

Why? Title Case Is So Much More Awesome :sob:

rtyler commented:
@daniel-beck  Be more :de:  :P

orrc committed:
Improve documentation for number of agent executors.

orrc committed:
Remove some relative links from the agent configuration page.

These help files can be shown at two different levels in the URL hierarchy:
- /computer/createItem
- /computer/<name>/configure

Therefore a relative URL will not work in all cases.  To link to the desired
pages correctly would probably require using Jelly or code.

orrc commented:
I'm declaring this particular PR as done.

The documentation for the individual _Launch methods_ could do with a rewrite, but I don't have the knowledge or time to definitively describe how things like JNLP tunnelling works :)

daniel-beck commented on war/src/main/webapp/help/system-config/master-slave/availability.html:
```diff
@@ -1,46 +1,56 @@
-                <li>They can execute on this slave</li>
+                <li>They have been in the queue for at least the specified <i>In
+                    demand delay</i> time period</li>
+                <li>They can be executed by this agent (e.g. have a matching
```
`e.g.` should be `i.e.`

orrc replied:
I believe I used `e.g.`, as there are other criteria which determine that the agent can be used.

For example, if there's no label expression, but "Use this node as much as possible" is set.  Plus I suppose other plugins could veto execution, even if the label expression did match.

daniel-beck commented on war/src/main/webapp/help/system-config/master-slave/description.html:
```diff
@@ -1,11 +1,6 @@
-  Optional human-readable description of this slave. This information
-  is displayed on the project configuration screen.
-
+  Optional human-readable description for this agent.
```
Unlike any other descriptions in Jenkins, it's not markup-formatted and not shown separately -- instead, it's displayed after the agent name, and should therefore be pretty short. Not sure whether relevant.

orrc replied:
Not sure. It doesn't really make a difference to the UI — I have a few agents with longer descriptions like "Mac Mini, OS X 10.11, 4-core, 32GB, Düsseldorf Datacentre" and nothing breaks :)

daniel-beck commented on war/src/main/webapp/help/system-config/nodeEnvironmentVariables.html:
```diff
@@ -1,4 +1,27 @@
+  alphabetical order of the "extra" part  of the name.
+  <p>
+  If the <i>Value</i> is empty or whitespace-only, it will not be added to the
+  environment.
```
WDYT? *… so this cannot be used to reset or unset an environment variable.*

orrc replied:
New:
> If the Value is empty or whitespace-only, it will not be added to the environment, and will not override or unset any existing environment variable with the same name that may already exist (e.g. defined by the system).

orrc replied:
(reworded it again)

daniel-beck commented:
Some minor suggestions, otherwise :+1: :+1: :+1: :+1: :+1:

orrc committed:
Add note about empty environment variables not overriding existing ones.

orrc commented:
I'll submit an ICLA tomorrow.

daniel-beck commented:
re-:+1:

orrc commented:
ICLA from me is here: https://github.com/jenkinsci/infra-cla/pull/30

daniel-beck commented on core/src/main/resources/hudson/slaves/EnvironmentVariablesNodeProperty/config.jelly:
```diff
@@ -24,13 +24,13 @@ THE SOFTWARE.
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define"
 	xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <f:entry title="${%List of key-value pairs}" help="${descriptor.getHelpPage()}">
+  <f:entry title="${%List of variables}" help="${descriptor.getHelpPage()}">
```
Is renaming these a great idea rather than fixing it in the `properties` file? This breaks all translations.

daniel-beck commented on core/src/main/resources/hudson/slaves/SimpleScheduledRetentionStrategy/config.jelly:
```diff
@@ -35,7 +35,7 @@ THE SOFTWARE.
                checkMessage="${%Scheduled Uptime is mandatory and must be a number.}"/>
     </f:entry>
-    <f:entry title="${%Keep on-line while jobs are running}">
+    <f:entry title="${%Keep online while jobs are running}"  help="/help/system-config/master-slave/demand/keepUpWhenActive.html">
```
Likewise re breaking translations

daniel-beck commented:
@orrc Could you take a quick look at my previous two comments? Maybe this can simply be resolved via `.properties` file. Otherwise, if we have translations, could you fix them so they don't break?

orrc commented:
@daniel-beck Good point. I'll be able to take a look this evening.

orrc committed:
Update translation keys for EnvironmentVariablesNodeProperty config.

orrc committed:
Update translation keys for SimpleScheduledRetentionStrategy config.

orrc commented:
@daniel-beck I updated the translations so that the keys match the new English text.
